3_time_series/Statistical_analysis.md

```diff
@@ -0,0 +1,749 @@
+from statsmodels.tsa.statespace.sarimax import SARIMAX
+from statsmodels.graphics.tsaplots import plot_acf, plot_pacf
+from statsmodels.tsa.stattools import adfuller
+from statsmodels.tsa.api import ExponentialSmoothing
+from time import time
+from dateutil.relativedelta import relativedelta
+import math
+from sklearn.metrics import mean_squared_error, mean_absolute_error
+from sklearn.linear_model import LinearRegression
+
+from scalecast.Forecaster import Forecaster
+from scalecast.auxmodels import auto_arima
+```
+
+```python
+sns.set_theme()
+%matplotlib inline
+```
+
+```python
+df = pd.read_csv('../data/AirPassengers.csv', parse_dates=[0], index_col=0,  
+                            date_parser=lambda s:datetime.strptime(s, '%Y-%m'))
+df = df.rename(columns={"#Passengers": "Num_Passengers"})
+df = df.asfreq(pd.infer_freq(df.index))
+```
+
+```python
+df.head()
+```
+
+```python
+print(df.dtypes)
+```
+
+```python
+print("Nan by columns:")
+for col in df.columns:
+    print(col, " - ", df[col].isna().sum())
+    
+print("Minimum number of passagers overall:", df["Num_Passengers"].min())
+print("Maximum number of passagers overall:", df["Num_Passengers"].max())
+print("Amount of measurements:", df["Num_Passengers"].count())
+```
+
+```python
+plt.title("Amount of air passagers by month")
+sns.lineplot(x="Month", y="Num_Passengers",
+             data=df)
+plt.show()
+```
+
+<!-- #region jp-MarkdownHeadingCollapsed=true tags=[] -->
+# STL Decompozition
+<!-- #endregion -->
+
+```python
+stl = STL(df)
+result = stl.fit()
+```
+
+```python
+seasonal, trend, resid = result.seasonal, result.trend, result.resid
+```
+
+```python
+plt.figure(figsize=(8,6))
+
+plt.subplot(4,1,1)
+plt.plot(df)
+plt.title('Original Series', fontsize=16)
+
+plt.subplot(4,1,2)
+plt.plot(trend)
+plt.title('Trend', fontsize=16)
+
+plt.subplot(4,1,3)
+plt.plot(seasonal)
+plt.title('Seasonal', fontsize=16)
+
+plt.subplot(4,1,4)
+plt.plot(resid)
+plt.title('Residual', fontsize=16)
+
+plt.tight_layout()
+```
+
+<!-- #region tags=[] jp-MarkdownHeadingCollapsed=true -->
+# ACF & PACF
+<!-- #endregion -->
+
+```python
+acf_plot = plot_acf(df.Num_Passengers, lags=20)
+```
+
+```python
+pacf_plot = plot_pacf(df.Num_Passengers, method='ywm')
+```
+
+<!-- #region jp-MarkdownHeadingCollapsed=true tags=[] -->
+# Stationarity
+<!-- #endregion -->
+
+```python
+def adf_test(timeseries):
+    print("Results of Dickey-Fuller Test:")
+    dftest = adfuller(timeseries, autolag="AIC")
+    dfoutput = pd.Series(
+        dftest[0:4],
+        index=[
+            "Test Statistic",
+            "p-value",
+            "#Lags Used",
+            "Number of Observations Used",
+        ],
+    )
+    print(dfoutput)
+    is_not = 'is' if dfoutput[1]<0.05 else 'is not'
+    print(f"Time series {is_not} stationary")
+```
+
+```python
+adf_test(df)
+```
+
+```python
+df.rolling(window = 12).mean().plot(figsize=(8,4), color="tab:red", title="Rolling Mean over 12 month period");
```

Comment:
sth like this helps visualization
`ax = df.plot()`
`df.rolling(window = 12).mean().plot(figsize=(8,4), color="tab:red", title="Rolling Mean over 12 month period", ax=ax);`